### ISSUE TITLE
Shadowed type var names break resolution in macros

### ISSUE THREAD

**Issue by KimBurgess** [crystal-lang/crystal#9488]
When a module uses a generic type var that shadows the name of an including types type var this appears to break the lookup / resolution of this.

```crystal
module Foo(A)
  macro included
    {{ pp A }}
  end
end

module Bar(T)
  macro included
    {{ pp T }}  # => Error: undefined constant T
  end
end

class Example(T)
  include Foo(Int32)
  include Bar(Int32)
end

Example(String).new
```
https://play.crystal-lang.org/#/r/9a0u

Reproducible on 0.35.0 and below.

**Comment by asterite** [comment]
What do you expect `T` to be there? Int32 or String?

**Comment by straight-shoota** [comment]
Initially, I thought it should probably be `String` because that's `T` in the scope where the macro hook is called. But if `Example` was not a generic type, it would clearly be `Int32`. And `Int32` also makes some sense because it's the lexiacal scope of the hook.
I suppose there's really no valid answer to this.

There are good reasons to ask for either one of them. So there's two solutions:

* Have some kind of disambiguation between generic arguments of module type and included type
* Disallow multiple generic arguments of the same name (but different type) for a type and its ancestors

**Comment by KimBurgess** [comment]
From a functionality point of view, having `T` resolve to `Int32` in the above example would be preferred (or at least preferred for my very narrow view of the world and current use cases).

From an adherence to current language semantics, resolving to `String` is likely more appropriate with the content that the included hook runs in.

Given the potential for ambiguity, disallowing this form of shadowing a provide a nice readable error explaining this seems like the neat option. e.g. Something similar to if you were to do:
```crystal
T = Int32
T = String
```

**Comment by asterite** [comment]
But there's no shadowing at all. The problem is wanting to access a type in a location where it's not clear where that type is.

I recommend coding whatever you want to do on a different way. Instead of including a module you can call a macro and pass the T that you want. The macro receives it, there's no. ambiguity.

I often feel that many of the features and bugs reported here stem from a very complex way of using the language. I believe things can be simpler.

**Comment by HertzDevil** [comment]
If there is a way to access `Bar(Int32)` inside the `included` hook then `T` can be recovered as `???.type_vars.first`. Does such a way exist?

**Comment by asterite** [comment]
I don't think so. The `inherited` and `included` macros lack an argument, which is the type that is being included. We could maybe it add it in a backwards-compatible way, I think.
